src/main/java/org/wso2/carbon/esb/connector/MQPublishQueue.java

```diff
@@ -0,0 +1,386 @@
+                    log.info("Reply received");
+                    MQMD md = new MQMD();
+                    md.copyFrom(message);
+                    message.getDataLength();
+                    log.info("Correlation ID " + new String(message.correlationId));
+                    log.info("Message ID " + new String(message.messageId));
+                    int strLen = message.getDataLength();
+                    byte[] strData = new byte[strLen];
+                    message.readFully(strData);
+                    log.info("Reply-\n" + new String(strData));
+                    Future<?> future;
+                    while (null == (future = listener[0])) {
+                        Thread.yield();
+                    }
+                    future.cancel(false);
+                    return;
+                } catch (MQException e) {
+                    log.info("Waiting for a reply");
+                } catch (IOException e) {
+
+                }
+
+            }, 0, config.getReplyTimeout(), TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Create report MQMesaage
+     */
+    MQMessage buildReportMessage(MQConfiguration config, String queueMessage) {
+
+        MQMessage mqMessage = new MQMessage();
+        mqMessage.messageId = config.getMessageID().getBytes();
+        mqMessage.correlationId = config.getCorrelationID().getBytes();
+        if (config.getPriority() != -1000) {
+            mqMessage.priority = config.getPriority();
+        }
+
+        int messageType = config.getMessageType();
+        mqMessage.messageType = messageType;
+        String reportQueue = config.getreplyQueue();
+        if (reportQueue != null) {
+            mqMessage.replyToQueueName = reportQueue;
+            mqMessage.report = CMQC.MQRO_COA_WITH_FULL_DATA | CMQC.MQRO_EXCEPTION_WITH_FULL_DATA | CMQC.MQRO_COD_WITH_FULL_DATA | CMQC.MQRO_EXPIRATION_WITH_FULL_DATA;
+            MQQueue reply;
+            reply = setReplyQueue(this.connectionBuilder, this.config);
+            if (reply != null) {
+                getReportMessage(reply, this.config);
+            }
+        } else {
+            log.info("Reply queue not specified");
+        }
+
+        if (config.isPersistent()) {
+            mqMessage.persistence = MQPER_PERSISTENT;
+        } else {
+            mqMessage.persistence = MQPER_NOT_PERSISTENT;
+        }
+        if (config.getgroupID() != null) {
+            mqMessage.groupId = config.getgroupID().getBytes();
+        }
+
+        GregorianCalendar cal = new GregorianCalendar();
+        cal.setTime(new Date(System.currentTimeMillis()));
+        mqMessage.putDateTime = cal;
+        try {
+            mqMessage.writeString(queueMessage);
+        } catch (Exception e) {
+            log.info("Error creating mq message" + e);
+        }
+        return mqMessage;
+    }
+
+    /**
+     * Get report message
+     */
+    void getReportMessage(final MQQueue replyQueue, MQConfiguration config) {
```

Comment:
I think can get rid of this method by pointing, ibm-mq inbound to reply queue. Handling Listening part a connector in not good IMO.